rust/src/shards/gui/widgets/image.rs

```diff
@@ -0,0 +1,180 @@
+    .into(),];
+}
+
+impl Default for Image {
+  fn default() -> Self {
+    let mut parents = ParamVar::default();
+    parents.set_name(PARENTS_UI_NAME);
+    Self {
+      parents,
+      requiring: Vec::new(),
+      scale: ParamVar::new((1.0, 1.0).into()),
+      texture: None,
+    }
+  }
+}
+
+impl Shard for Image {
+  fn registerName() -> &'static str
+  where
+    Self: Sized,
+  {
+    cstr!("UI.Image")
+  }
+
+  fn hash() -> u32
+  where
+    Self: Sized,
+  {
+    compile_time_crc32::crc32!("UI.Image-rust-0x20200101")
+  }
+
+  fn name(&mut self) -> &str {
+    "UI.Image"
+  }
+
+  fn help(&mut self) -> OptionalString {
+    OptionalString(shccstr!("Display an image in the UI."))
+  }
+
+  fn inputTypes(&mut self) -> &Types {
+    &IMAGE_TYPES
+  }
+
+  fn inputHelp(&mut self) -> OptionalString {
+    OptionalString(shccstr!("The image to display"))
+  }
+
+  fn outputTypes(&mut self) -> &Types {
+    &IMAGE_TYPES
+  }
+
+  fn outputHelp(&mut self) -> OptionalString {
+    OptionalString(shccstr!("The output of this shard will be its input."))
+  }
+
+  fn parameters(&mut self) -> Option<&Parameters> {
+    Some(&IMAGE_PARAMETERS)
+  }
+
+  fn setParam(&mut self, index: i32, value: &Var) -> Result<(), &str> {
+    match index {
+      0 => Ok(self.scale.set_param(value)),
+      _ => Err("Invalid parameter index"),
+    }
+  }
+
+  fn getParam(&mut self, index: i32) -> Var {
+    match index {
+      0 => self.scale.get_param(),
+      _ => Var::default(),
+    }
+  }
+
+  fn requiredVariables(&mut self) -> Option<&ExposedTypes> {
+    self.requiring.clear();
+
+    // Add UI.Parents to the list of required variables
+    util::require_parents(&mut self.requiring, &self.parents);
+
+    Some(&self.requiring)
+  }
+
+  fn warmup(&mut self, context: &Context) -> Result<(), &str> {
+    self.parents.warmup(context);
+    self.scale.warmup(context);
+
+    Ok(())
+  }
+
+  fn cleanup(&mut self) -> Result<(), &str> {
+    self.scale.cleanup();
+    self.parents.cleanup();
+
+    Ok(())
+  }
+
+  fn activate(&mut self, _context: &Context, input: &Var) -> Result<Var, &str> {
+    if let Some(ui) = util::get_current_parent(*self.parents.get())? {
+      let texture = &*self.texture.get_or_insert_with(|| {
+        let image: SHImage = input.try_into().unwrap();
+        let image: egui::ColorImage = (&image).into();
+
+        ui.ctx().load_texture("example", image) // FIXME name
+      });
+
+      let scale: (f32, f32) = self.scale.get().try_into()?;
+      let scale: egui::Vec2 = scale.into();
+      ui.image(texture, scale * texture.size_vec2());
+
+      Ok(*input)
+    } else {
+      Err("No UI parent")
+    }
+  }
+}
+
+impl TryFrom<&Var> for SHImage {
+  type Error = &'static str;
+
+  #[inline(always)]
+  fn try_from(var: &Var) -> Result<Self, Self::Error> {
+    if var.valueType != SHType_Image {
+      Err("Expected Image variable, but casting failed.")
+    } else {
+      unsafe { Ok(var.payload.__bindgen_anon_1.imageValue) }
```

Comment:
Maybe helper

Reply:
That's already a helper. But I can move it to `types.rs`.